### PR TITLE
implement NaCl - phase 1

### PR DESCRIPTION
**Pull request by James-Pickett**
We are moving krypto toward implementing [NaCl](https://doc.libsodium.org/) since it simplifies the boxing process and there is less room for error. However, in order to have future compatibility with MacOS secure enclave, we are pre-computing P256 elliptic curve keys and passing them to NaCl rather than using the native NaCl ed25519 elliptic curve keys.

Below is a rough outline of how I plan to move forward with this and I'll do my best to keep PRs smallish.

* phase 1 - basic cross language NaCl seal / open
* phase 2 - encode sealed NaCl boxes into pngs
* phase 3- integrate TPM keys
* phase 4 - add sign / verify ?
* phase 5 - integrate Secure Enclave ?

### CODE REVIEW

**Review by directionless — COMMENTED** [review]
Seems pretty reasonable. Though I think we should consider `[]byte` vs string. 

For signing, I wonder if that's _not_ part of nacler, and instead functions on the keyer side

File: pkg/nacler/keyers/localecdsa/localecdsa.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	key *ecdsa.PrivateKey
+}
+
+func New(key *ecdsa.PrivateKey) *localecdsa {
+	return &localecdsa{
+		key: key,
+	}
+}
+
+func (l *localecdsa) SharedKey(counterParty ecdsa.PublicKey) ([32]byte, error) {
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply pkg/nacler/keyers/localecdsa/localecdsa.go:18]: I learned it from here https://asecuritysite.com/ecc/goecdh, adding to comments

File: pkg/nacler/nacler.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	"crypto/ecdsa"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"io"
+
+	"golang.org/x/crypto/nacl/box"
+)
+
+type Keyer interface {
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply pkg/nacler/nacler.go:13]: nope, will unexport

File: pkg/nacler/nacler.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	SharedKey(counterParty ecdsa.PublicKey) ([32]byte, error)
+}
+
+type Nacler struct {
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply pkg/nacler/nacler.go:17]: yes, for cross tests, to be able to iterate of different implementations of different keyers

or we could push the different keyer tests down to package level

File: pkg/nacler/nacler.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func New(keyer Keyer, counterParty ecdsa.PublicKey) *Nacler {
+	return &Nacler{
+		keyer:        keyer,
+		counterParty: counterParty,
+	}
+}
+
+func (n *Nacler) Seal(plainText []byte) (string, error) {
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply pkg/nacler/nacler.go]: updated all the open / seal funcs to use []byte

File: lib/krypto/nacler.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      raise "Missing counter party" unless counterparty
+      @counterparty = counterparty
+    end
+
+    def seal(plaintext)
+      shared = @key.dh_compute_key(@counterparty)
+      shared_key = OpenSSL::Digest::SHA256.digest(shared)
+      box = RbNaCl::SimpleBox.from_secret_key(shared_key)
+      ciphertext = box.encrypt(plaintext)
+      Base64.strict_encode64(ciphertext)
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply lib/krypto/nacler.rb]: dropped all the b64ing

File: test/krypto/nacler_test.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+  ALICENACLER = Krypto::Nacler.new(ALICEKEY, BOBKEY.public_key)
+  BOBNACLER = Krypto::Nacler.new(BOBKEY, ALICEKEY.public_key)
+
+  MESSAGE_TO_SEAL = "this is the plaintext of the sealed message"
+
+  define_method(:test_alice_seal_bob_open) do
+    cipertext = ALICENACLER.seal(MESSAGE_TO_SEAL)
+    assert_equal(MESSAGE_TO_SEAL, BOBNACLER.open(cipertext))
+  end
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply test/krypto/nacler_test.rb]: done

File: cross_language_tests/nacler.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+cmd = args.shift
+testcase = MessagePack.unpack(Base64.strict_decode64(File.read(args.shift)))
+
+key = OpenSSL::PKey::EC.new(testcase["Key"])
+counterparty = OpenSSL::PKey::EC.new(testcase["Counterparty"])
+nacler = Krypto::Nacler.new(key, counterparty.public_key)
+
+case cmd
+when "seal"
+  puts(nacler.seal(testcase["Plaintext"]))
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply cross_language_tests/nacler.rb:19]: i did not need to update this

File: cross_language_tests/nacler_cross_test.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+		//TODO: add tpm keyer
+	}
+
+	for _, aliceNacler := range naclers {
+		aliceNacler := aliceNacler
+
+		for _, messageToSeal := range testMessages {
+			messageToSeal := messageToSeal
+
+			t.Run("Alice seals in go, Bob opens in ruby", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply cross_language_tests/nacler_cross_test.go]: good catch, updated to show what keyer it's using

File: pkg/nacler/nacler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (directionless) [reply pkg/nacler/nacler.go:16]: Eh. Easier just to export it.

File: lib/krypto/nacler.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+      raise "Missing key" unless key
+      @key = key
+      raise "Missing counter party" unless counterparty
+      @counterparty = counterparty
+    end
+
+    def seal(plaintext)
+      shared = @key.dh_compute_key(@counterparty)
+      shared_key = OpenSSL::Digest::SHA256.digest(shared)
+      box = RbNaCl::SimpleBox.from_secret_key(shared_key)
```
Comment (directionless) [inline lib/krypto/nacler.rb:18]: If these don't change, they could be in `initialize` and not seal. But at the volume here, it doesn't much matter. (The performance overhead seems fine)

Though it does mean repeating it between seal and open.

File: pkg/nacler/nacler_test.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"testing"
+
+	"github.com/kolide/krypto/pkg/nacler/keyers/localecdsa"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNacler(t *testing.T) {
```
Comment (directionless) [inline pkg/nacler/nacler_test.go:13]: You could add the failure case -- Mallory can't unseal

**Review by directionless — APPROVED** [review]
I'd take it

File: pkg/nacler/nacler_test.go (hunk 10 above)
Comment (James-Pickett) [reply pkg/nacler/nacler_test.go:13]: I'll add this in the next PR!

File: lib/krypto/nacler.rb (hunk 9 above)
Comment (James-Pickett) [reply lib/krypto/nacler.rb:18]: I'll add that to the next PR!

**Pull request merged**